File: 02 Contrast Adjustment/histeq.py
import numpy as np
import matplotlib.pyplot as plt
import cv2
'''
1. Find frequency of each pixel.
2. Find the probability density fucntion of each frequency.
3. Find the cumulative histogram of each pixel.
4. Find the cumulative distribution probability of each pixel.
5. Calculating final new pixel value by cdf * (no of bits).
6. Replace it with oroginal values.
'''

img = cv2.imread('./img/Camera Man.bmp', cv2.IMREAD_GRAYSCALE)
cv2.imshow('src',img)
cv2.waitKey(0)


plt.subplot(2,2,1)
plt.imshow(img, cmap='gray')



#To display image before equalization




a = np.zeros((256,),dtype=np.float16)
b = np.zeros((256,),dtype=np.float16)

height,width=img.shape

#finding histogram
for i in range(width):
    for j in range(height):
        g = img[j,i]
        a[g] = a[g]+1

#print(a)
plt.subplot(2,2,2)
plt.plot(a)

#performing histogram equalization
tmp = 1.0/(height*width)
b = np.zeros((256,),dtype=np.float16)

for i in range(256):
    for j in range(i+1):
        b[i] += a[j] * tmp;
    b[i] = round(b[i] * 255);

# b now contains the equalized histogram
b=b.astype(np.uint8)
plt.subplot(2,2,4)
plt.plot(b)
#print(b)

#Re-map values from equalized histogram into the image
for i in range(width):
    for j in range(height):
        g = img[j,i]
        img[j,i]= b[g]


plt.subplot(2,2,3)
plt.imshow(img, cmap='gray')

cv2.imshow('dest',img)
cv2.waitKey(0)
cv2.destroyAllWindows()

# plt.show()
def histeq(img):
    height, width = img.shape
    intensities = np.linspace(start=0, stop=255, num=256, endpoint=True, dtype=int)
    img = img.flatten()

    # print(len(img)) # 262144
    frequencies = np.zeros(256, dtype=np.uint8)
    # frequencies = np.zeros(256, dtype=np.uint8)
    print(frequencies)
    for pixel in range(len(img)):
        frequencies[img[pixel]] = frequencies[img[pixel]] + 1
    print(frequencies)

File: 02 Contrast Adjustment/histeq.m
I = imread('D:\2020.2.26\ComputerEngineeringUni\6 Fundamentals Of Computer Vision\FUM 98-02\Homeworks\Images\2\Camera Man.bmp');
[R,C]=size(I);
n = R*C;
Hist=zeros(256,1);
pdf = zeros(256,1);
cum = zeros(256,1);
cdf = zeros(256,1);
out = zeros(256,1);
equlizedHist = zeros(256,1);
equlizedImg = uint8(zeros(R,C));

for r = 1:R  
    for c=1:C         
        Hist(img(r,c)+1,1)=Hist(img(r,c)+1,1)+1;
        pdf(img(r,c)+1,1) = Hist(img(r,c)+1,1)/n;
    end
end


sum = 0;

for i = 1:size(pdf)
    sum = sum + Hist(i,1);
    cum(i,1) = sum;
    cdf(i,1) = cum(i,1)/n;
    out(i,1) = round(cdf(i,1) * 255);
end

for r = 1:R
    for c = 1:C
        equlizedImg(r,c) = out(I(r,c) + 1,1);
        equlizedHist(equlizedImg(r,c)+1,1)=equlizedHist(equlizedImg(r,c)+1,1)+1;
    end
end


subplot(2,2,1);
imshow(img);
title('Camera Man');

subplot(2,2,2);
bar(Hist);
title('Camera Man Histogram');

subplot(2,2,3);
imshow(equlizedImg);
title('Equalized Image Of Camera Man ');

subplot(2,2,4);
bar(equlizedHist);
title('Equalized Histogram Of Camera Man ');




